ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/BulkSync.hs

```diff
@@ -0,0 +1,613 @@
+  currentChain
+  fetchedBlocks
+  fetchedMaxSlotNo
+  chainSelStarvation
+  ( peersOrder0,
+    writePeersOrder,
+    demoteCSJDynamoAndIgnoreInflightBlocks
+    )
+  candidatesAndPeers = do
+    peersOrder <-
+      peersOrder0
+        -- Align the peers order with the actual peers; this consists in removing
+        -- all peers from the peers order that are not in the actual peers list and
+        -- adding at the end of the peers order all the actual peers that were not
+        -- there before.
+        & alignPeersOrderWithActualPeers
+          (map (\(_, (_, _, _, peer, _)) -> peer) candidatesAndPeers)
+        -- If the chain selection has been starved recently, that is after the
+        -- current peer started (and a grace period), then the current peer is
+        -- bad. We push it at the end of the queue, demote it from CSJ dynamo,
+        -- and ignore its in-flight blocks for the future.
+        & checkLastChainSelStarvation
+
+    -- Compute the actual block fetch decision. This contains only declines and
+    -- at most one request. 'theDecision' is therefore a 'Maybe'.
+    let (theDecision, declines) =
+          fetchDecisionsBulkSync
+            fetchDecisionPolicy
+            currentChain
+            fetchedBlocks
+            fetchedMaxSlotNo
+            peersOrder
+            candidatesAndPeers
+
+    -- If the peer that is supposed to fetch the block is not the current one in
+    -- the peers order, then we have shifted our focus: we make the new peer our
+    -- current one and we put back the previous current peer at the beginning of
+    -- the queue; not the end, because it has not done anything wrong.
+    maybeSetCurrentPeer theDecision peersOrder
+
+    pure $
+      map (first Right) (maybeToList theDecision)
+        ++ map (first Left) declines
+    where
+      alignPeersOrderWithActualPeers :: [peer] -> PeersOrder peer -> PeersOrder peer
+      alignPeersOrderWithActualPeers
+        actualPeers
+        PeersOrder {peersOrderCurrent, peersOrderStart, peersOrderOthers} =
+          let peersOrderCurrent' = case peersOrderCurrent of
+                Just peersOrderCurrent_ | peersOrderCurrent_ `elem` actualPeers -> peersOrderCurrent
+                _ -> Nothing
+              peersOrderOthers' =
+                filter (`elem` actualPeers) peersOrderOthers
+                  ++ filter (\peer -> peer `notElem` peersOrderOthers && Just peer /= peersOrderCurrent) actualPeers
+           in PeersOrder
+                { peersOrderCurrent = peersOrderCurrent',
+                  peersOrderOthers = peersOrderOthers',
+                  peersOrderStart
+                }
+
+      checkLastChainSelStarvation :: PeersOrder peer -> m (PeersOrder peer)
+      checkLastChainSelStarvation
+        peersOrder@PeersOrder {peersOrderCurrent, peersOrderStart, peersOrderOthers} = do
+          lastStarvationTime <- case chainSelStarvation of
+            ChainSelStarvationEndedAt time -> pure time
```

Comment:
I'm not finding any place in the repos where `ChainSelStarvationEndedAt` is created. Is this value produced ever?

Reply:
It should be created by `getChainSelMessage` in `ouroboros-consensus`. See the [current state of this function](https://github.com/IntersectMBO/ouroboros-consensus/blob/0c8ef3713343960f2e91b826464be69346d932b0/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs#L519-L548)

Reply:
Note: at the point where you said this, as we discussed on Slack, it was indeed missing. I think the new version is correct, but it doesn't look like it from the logs, so please have a close look at it!